tensorboard/lazy.py

```diff
@@ -19,36 +19,63 @@
 import types
 
 
-class LazyLoader(types.ModuleType):
-  """Lazily import a module, mainly to avoid pulling in large dependencies."""
+def lazy_load(name):
+  """Decorator to define a function that lazily loads the module 'name'.
 
-  # The lint error here is incorrect.
-  def __init__(self, local_name, parent_module_globals, name):  # pylint: disable=super-on-old-class
-    self._local_name = local_name
-    self._parent_module_globals = parent_module_globals
+  This can be used to defer importing troublesome dependencies - e.g. ones that
+  are large and infrequently used, or that cause a dependency cycle -
+  until they are actually used.
 
-    super(LazyLoader, self).__init__(name)
+  Args:
+    name: the fully-qualified name of the module; typically the last segment
+      of 'name' matches the name of the decorated function
 
-  def _load(self):
-    # Import the target module and insert it into the parent's namespace
-    module = importlib.import_module(self.__name__)
-    self._parent_module_globals[self._local_name] = module
+  Returns:
+    Decorator function that produces a lazy-loading module 'name' backed by the
+    underlying decorated function.
+  """
+  def wrapper(load_fn):
+    # Wrap load_fn to call it exactly once and update __dict__ afterwards to
+    # make future lookups efficient (only failed lookups call __getattr__).
+    @_return_once
+    def load_once(self):
+      module = load_fn()
+      self.__dict__.update(module.__dict__)
+      return module
 
-    # Update this object's dict so that if someone keeps a reference to the
-    #   LazyLoader, lookups are efficient (__getattr__ is only called on lookups
-    #   that fail).
-    self.__dict__.update(module.__dict__)
+    # Define a module that proxies getattr() and dir() to the result of calling
+    # load_once() the first time it's needed. The class is nested so we can close
+    # over load_once() and avoid polluting the module's attrs with our own state.
+    class LazyModule(types.ModuleType):
+      def __getattr__(self, attr_name):
+        return getattr(load_once(self), attr_name)
 
-    return module
+      def __dir__(self):
+        return dir(load_once(self))
 
-  def __getattr__(self, item):
-    module = self._load()
-    return getattr(module, item)
+      def __repr__(self):
+        return '<module \'%s\' (LazyModule)>' % self.__name__
 
-  def __dir__(self):
-    module = self._load()
-    return dir(module)
+    return LazyModule(name)
+  return wrapper
+
+
+def _return_once(f):
+  """Decorator that calls f() once, then returns that value repeatedly."""
+  not_called = object()  # Unique "not yet called" sentinel object.
+  # Cache result indirectly via a list since closures can't reassign variables.
+  cache = [not_called]
+  lock = threading.Lock()
+  @functools.wraps(f)
+  def wrapper(*args, **kwargs):
+    if cache[0] == not_called:
+      with lock:
+        if cache[0] == not_called:
+          cache[0] = f(*args, **kwargs)
```

Comment:
This open call can deadlock if the `load_fn` references the very module
that it corresponds to:

```python
>>> from tensorboard import lazy
>>> @lazy.lazy_load("foo")
... def foo():
...   print("loading")
...   return foo.x
... 
>>> foo.y
loading
^C^C^C^\Quit
```

I can’t think of a valid reason to do this, so no change requested (if
you don’t mind, I’ll submit a follow-up PR to detect this and raise a
nice error message instead of hanging).

Reply:
Actually this is a good point, since this could also be triggered by an import cycle:

```python
@lazy.lazy_load("inner")
def inner():
  return outer.foo

@lazy.lazy_load("outer")
def outer():
  return inner
```

And I've had enough headaches with those without deadlock too :p  Changed to use a reentrant lock so you just get a recursion depth exceeded error instead, WDYT?